CompoundEditor : Allow editors to be linked to each other or the scene selection

themissingcow opened the pull request:
The pinning menu has been significantly re-vamped to support linking editors to one another, or the scene selection. See individual commits for details.

![image](https://user-images.githubusercontent.com/896779/61528050-bf2ddd80-aa15-11e9-8a72-bf5184b3d56b.png)

NOTE: These changes highlight an existing bug whereby we don't clear the highlight state of button images when its menu is closed.

## User Facing Changes

 - New pinning menu items to allow editors to be linked and follow scene selection.
 - Re-structured pinning menu bookmarks section
 - New pinning button icons to reflect the editors link state
 - Color decorations next to the pinning button for linked editors
   to help understand link relationships

## API Changes

 - Adds `GafferSceneUI.SceneSet`
 - `NodeSetEditor` :
   - Adds `setNodeSetDriver`/`getNodeSetDriver` to link editors.
   - Adds `drivenNodeSets` to query linked editors.
   - Adds `nodeSetDriverChangedSignal`, `nodeSetDriverChangedSignal` signals to allow link status to be observed.
   - Adds registerNodeSetDriverMode to allow custom link modes to be registered by other code.
 - `NodeSetEditorAlgo` provides extended query functions.

johnhaddon commented on python/GafferSceneUITest/SourceSetTest.py:
```diff
@@ -0,0 +1,210 @@
+		c = Gaffer.Context()
+
+		h = GafferSceneUI.SourceSet( c, n )
+		ca = h.memberAddedSignal().connect( added )
```
Not sure if you've come across it yet, or if you just don't think its that applicable here. But `GafferTest.CapturingSlot()` is very handy for testing signals - it simply accumulates a record of all the calls made to it.

themissingcow replied:
Ahh I hadn't seen that. It did seem like it's useful in this case to check the actual objects passed too and the resulting state accumulation? What do you think?

johnhaddon replied:
I think you could argue that `CapturingSlot` would be better here. The current tests don't care what order the signals are emitted in, or how many times, just that they eventually agree with what we want. If we used CaptuingSlot, we'd be asserting that they were only triggered once per edit, and that remove was before add. Not a big deal though - I just wanted to make you aware of CapturingSlot, since it is frequently useful.

johnhaddon commented on python/GafferUI/NodeSetEditorAlgo.py:
```diff
@@ -0,0 +1,69 @@
+
+import GafferUI
+
+def drivingEditor( editor ) :
```
Do you think this could be `EditorAlgo` rather than `NodeSetEditorAlgo`? I like the use of Algo to keep the classes a bit more minimal, but I'm wary of having too many scattered algo namespaces. My hope would be that in the future we'll find more citizens for `EditorAlgo` than we would for `NodeSetEditorAlgo` alone...

themissingcow replied:
Yeah good point. Also if you consider the point about `NodeSetEditor` ending up using it (which had happened as part of a bug-fix) it doesn't make much sense. I went for just moving it into `NodeSetEditor` in the end, as it makes most call sites neater and easier to read anyway. There would have also only been one short method in the end... https://github.com/GafferHQ/gaffer/pull/3270/commits/7548da0d9d0b504dc4e99163cd5bb1197efd54fd

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
-	# added to and removed from the set, the UI will be updated automatically
-	# to show them. This also calls `nodeSet.setRemoveOrphans( True )` so that
-	# deleted nodes are not visible in the UI.
+	## Sets the nodes that will be displayed by this editor.
```
Restore original formatting.

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/0f4055a0353ab13b52dd2c1589ce6c6d069d85a1

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+	# NOTE: The driver will be updated *after* the node set, such that calling
+	# `getNodeSetDriver` in the nodeSetChangedSignal will return the departing
+	# driver. If the listener requires accurate information about the driver, then
+	# the nodeSetDriverChanegd signal should be used instead.
```
nodeSetDriverChanegd > nodeSetDriverChanged

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/0f4055a0353ab13b52dd2c1589ce6c6d069d85a1

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+	# )` so that deleted nodes are not visible in the UI.
+	#
+	# This will break any editor links, if set.
+	# NOTE: The driver will be updated *after* the node set, such that calling
```
Is there a good reason for this? It seems unexpected - as per my comment for the SourceSet signals, we want to emit signals only when we're in a consistent state from an observer's perspective.

themissingcow replied:
Reflected our chat off-line in https://github.com/GafferHQ/gaffer/pull/3270/commits/318060f583e5e0e266c3d592c86800da5bc25fc0

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+		if drivingEditor is not None :
+			assert( isinstance( drivingEditor, GafferUI.NodeSetEditor ) )
+			# We also need to stop people creating infinite loops
+			if GafferUI.NodeSetEditorAlgo.drivesNodeSet( self, drivingEditor ) :
```
This seems odd - the general idea with Algo is that it provides functionality above and beyond that of the class itself. To have the class depend on the algo is a bit backwards, no?

themissingcow replied:
Totally! Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/7548da0d9d0b504dc4e99163cd5bb1197efd54fd

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+
+		if self.__nodeSetDriver :
+			previousDriver = self.__nodeSetDriver["weakDriver"]()
+			# It may have died, so we sill will have data, but the ref wil be dead
```
sill -> still?

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/0f4055a0353ab13b52dd2c1589ce6c6d069d85a1

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+					# We need to be able to track un-pinning, so we need to see if someone
+					# else has changed the nodeSet of the editor by some other means, so
+					# keep a copy here.
+					self.__nodeSetDriver["lastNodeSet"] = nodeSet
```
I can't find any other reference to this - is it doing anything?

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/d0aa0e2d70aaae24dd0d459b5fc99220291c8da7

johnhaddon commented on python/GafferUI/NodeSetEditor.py:
```diff
@@ -61,22 +77,149 @@ def __init__( self, topLevelWidget, scriptNode, **kw ) :
+		if recurse :
+			distant = {}
+			for editor in driven :
+				distant.update( editor.drivenNodeSets( recurse = True ) )
```
Is this two-stage update using `distant` any different than just calling `driven.update()` directly?

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3270/commits/07eba8b8f359b989d1f79a1a1c7a72a189cc57eb

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -543,37 +562,89 @@ def __updatePinningButton( self, *unused ) :
+		# ScriptNode selection are considered 'pinned'
+		return True
+
+	def __pinToGraphEditorSelection( self, *args ) :
```
GraphEditor -> Script? The GraphEditor doesn't own the selection, and we want the CompoundEditor to have as little knowledge of individual editors as possible.

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/dacd021d21f96b8465056a36b6f92ff76dc1b658

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -543,37 +562,89 @@ def __updatePinningButton( self, *unused ) :
+		editor = self.getCurrent()
+		editor.setNodeSet( Gaffer.StandardSet( list( editor.scriptNode().selection() ) ) )
+
+	def __followGraphEditorSelection( self, *args ) :
```
GraphEditor -> Script?

themissingcow replied:
Fixed inhttps://github.com/GafferHQ/gaffer/pull/3270/commits/dacd021d21f96b8465056a36b6f92ff76dc1b658

johnhaddon commented on startup/gui/editor.py:
```diff
@@ -44,8 +45,106 @@ def __editorCreated( editor ) :
+
+#### Editor follows Scene Selection
+
+# This makes the supplied editor follow the source node of the scene selectionf another editor.
```
selectionf -> selection of

johnhaddon commented on startup/gui/editor.py:
```diff
@@ -44,8 +45,106 @@ def __editorCreated( editor ) :
+
+### Pinning Menu Items
+
+# Helper for registering unique menu items considering curent state
```
curent -> current

johnhaddon commented on startup/gui/editor.py:
```diff
@@ -44,8 +45,106 @@ def __editorCreated( editor ) :
+		return
+
+	# Generally, we consider the Viewer/Hierarchy views 'focal' editors as
+	# they can addect the context selection/expansion state, etc...
```
addect -> affect

themissingcow commented on include/GafferSceneUI/SourceSet.h:
```diff
@@ -0,0 +1,116 @@
+namespace GafferSceneUI
+{
+
+/// The SourceSet provides a Set implementation that adjusts it's membership such that
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/0f4055a0353ab13b52dd2c1589ce6c6d069d85a1

themissingcow commented on include/GafferSceneUI/SourceSet.h:
```diff
@@ -0,0 +1,116 @@
+
+/// The SourceSet provides a Set implementation that adjusts it's membership such that
+/// it contains the source node for the current scene selection, given a context and
+/// a traget node from which to observe the scene.
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/0f4055a0353ab13b52dd2c1589ce6c6d069d85a1

themissingcow commented on include/GafferSceneUI/SourceSet.h:
```diff
@@ -0,0 +1,116 @@
+
+	public :
+
+		SourceSet( Gaffer::ContextPtr context, Gaffer::StandardSetPtr nodeSet );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/891eb75444f4b32dbb8947d8f6db495d98e15e4b

themissingcow commented on include/GafferSceneUI/SourceSet.h:
```diff
@@ -0,0 +1,116 @@
+		Gaffer::ContextPtr getContext() const;
+
+		void setNodeSet( Gaffer::StandardSetPtr nodeSet );
+		Gaffer::StandardSetPtr getNodeSet() const;
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/891eb75444f4b32dbb8947d8f6db495d98e15e4b

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+		// We want the last selected node rather than the first
+		for( size_t i = m_nodes->size(); i > 0; --i )
+		{
+			if( Node *node = dynamic_cast<Node*>( m_nodes->member( i - 1 ) ) )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/891eb75444f4b32dbb8947d8f6db495d98e15e4b

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+	if( m_nodes && m_nodes->size() > 0 )
+	{
+		// We want the last selected node rather than the first
+		for( size_t i = m_nodes->size(); i > 0; --i )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/ebccfc0537be325d4bdb7cec612e24ddb422cc34

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+		{
+			if( ScenePlug *sourcePlug = SceneAlgo::source( m_scenePlug.get(), path ) )
+			{
+				newSourceNode = dynamic_cast<SceneNode*>( sourcePlug->node() );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/ebccfc0537be325d4bdb7cec612e24ddb422cc34 via https://github.com/GafferHQ/gaffer/pull/3270/commits/db51515fa39fe2d9fd6de86fdf42d4bfa44ba8fa

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+		const ScenePlug::ScenePath path = ContextAlgo::getLastSelectedPath( m_context.get() );
+		if( !path.empty() && SceneAlgo::exists( m_scenePlug.get(), path ) )
+		{
+			if( ScenePlug *sourcePlug = SceneAlgo::source( m_scenePlug.get(), path ) )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/055cc20abf246e46d883a95177a8bca7c03586bd

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+		// If we dont' have a valid context or scene plug, but do have a node,
+		// just assume thats the source. This generally makes sense for non-
+		// scene nodes.
+		newSourceNode = dynamic_cast<Node*>( m_nodes->member( m_nodes->size() - 1 ) );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/891eb75444f4b32dbb8947d8f6db495d98e15e4b

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+	else if( m_nodes && m_nodes->size() > 0 )
+	{
+		// If we dont' have a valid context or scene plug, but do have a node,
+		// just assume thats the source. This generally makes sense for non-
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/0f4055a0353ab13b52dd2c1589ce6c6d069d85a1

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+	{
+		if( m_sourceNode )
+		{
+			memberRemovedSignal()( this, m_sourceNode.get() );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/dec7dca3b09721dbb7d0130a32982a1452b8a3f5

themissingcow commented on src/GafferSceneUI/SourceSet.cpp:
```diff
@@ -0,0 +1,227 @@
+		return oIt->get();
+	}
+
+	InputScenePlugIterator iIt( node );
```
In a previous WIP version, it hadn't been correctly finding source nodes for InteractiveRender nodes, etc... Checking with the final implementation, it's working with those nodes and the test still pass that consider other node types, so must have been some other problem between the chair and the keyboard. Removed in https://github.com/GafferHQ/gaffer/pull/3270/commits/41ff3991901dc0b54764d869704cdc6ace66bbbf

themissingcow commented on src/GafferSceneUIModule/SourceSetBinding.cpp:
```diff
@@ -0,0 +1,60 @@
+
+void GafferSceneUIModule::bindSourceSet()
+{
+	{
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/891eb75444f4b32dbb8947d8f6db495d98e15e4b

themissingcow commented on src/GafferSceneUIModule/SourceSetBinding.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+	{
+		boost::python::scope s = IECorePython::RunTimeTypedClass<SourceSet>()
+			.def( boost::python::init<ContextPtr, StandardSetPtr>() )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/891eb75444f4b32dbb8947d8f6db495d98e15e4b

themissingcow commented on src/GafferSceneUIModule/SourceSetBinding.cpp:
```diff
@@ -0,0 +1,60 @@
+	{
+		boost::python::scope s = IECorePython::RunTimeTypedClass<SourceSet>()
+			.def( boost::python::init<ContextPtr, StandardSetPtr>() )
+			.def( "setContext", &SourceSet::setContext )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/07400bdbb751a845bb0f7f7f86de5334391c72a6

themissingcow commented on python/GafferSceneUITest/SourceSetTest.py:
```diff
@@ -0,0 +1,210 @@
+		# Test scene selection changed
+
+		GafferSceneUI.ContextAlgo.setSelectedPaths( c, IECore.PathMatcher( [ a	 ] ) )
+		self.assertEqual( set( h ), set( [ s["pA"], ] ) )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/3270/commits/e79df26cd2a1d7f9cd96627014c7a94f39ec4a80

johnhaddon commented on include/Gaffer/MetadataAlgo.h:
```diff
@@ -91,6 +91,9 @@ GAFFER_API bool getReadOnly( const GraphComponent *graphComponent );
 GAFFER_API bool getChildNodesAreReadOnly( const Node *node );
 
+/// Looks for the most distant parent of node whose children are all read-only.
+GAFFER_API Node *ancestorWithReadOnlyChildNodes( Node *node );
```
Do you think there's a better way of phrasing this? One of the reasons I kept it private originally was that it sounded so awkward, and didn't even quite do what I wanted anyway. At both call sites we just want to know which node to show, `node` or some ancestor, but this doesn't do that, because it sometimes returns `null`. Perhaps `closestEditableAncestor()`?

themissingcow replied:
As per out-of-band discussion. We're confining this to private implementation for now, until we can figure out a sane way to adopt this in `MetadataAlgo`.

johnhaddon commented:
Thanks Tom, and sorry for the unbatched comments. I don't know how best to deal with replying to previous inline discussions at the same time as making new comments on new code. Clearly, not like this though.

Now I can run this, I'll do some testing here too. I'm a bit wary of rushing in such a big change to such a central mechanism just prior to release, but it does make a logical companion to the other layout improvements. I'll see if Murray has any time to do further testing. Would be great if @mattigruener and @andrewkaufman could do some too, and perhaps canvas user feedback?

themissingcow commented:
> Thanks Tom, and sorry for the unbatched comments. I don't know how best to deal with replying to previous inline discussions at the same time as making new comments on new code. Clearly, not like this though.

Sorry thats probably my fault, for some reason in my initial replying GitHub decided not to let me start a review... 

I've fixed the remaining scope/pointer issues in https://github.com/GafferHQ/gaffer/pull/3270/commits/2537eccc261dbce51c4e1ba3b877dcc1c69ac37c, and removed the commit that extended `MetadataAlgo` and migrated `SceneHistoryUI`.

johnhaddon commented:
Thanks Tom - all looking good so far. Could you do the rebase/squash we talked about and fix the TypeIds conflict?

medubelko commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -585,6 +656,33 @@ def __pinningButtonContextMenu( self, button ) :
+		# Signals use this to anchor insertions of actions or state items appropriately.
+		m.append( "/ActionsDivider", { "divider" : True } )
+
+		m.append( "/Re-pin Current Node Selection" if isPinned else "Pin Current Node Selection", {
```
How about:

`m.append( "/Pin to Current Node Selection", {`

I find the "re" prefix more confusing than anything. Am I pinning to the _same_ node as before? I don't see a reason to include the prefix. In every circumstance, I'm just pinning to whatever's currently selected. The "✓ Pinned" entry and icon should be enough to tell me that I'm already pinned to a node, instead of linked/following.

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/commit/aca4e28c8ebb8804581443491d1e9f96db3798b5

medubelko commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -585,6 +656,33 @@ def __pinningButtonContextMenu( self, button ) :
+		isPinned  = self.__editorIsPinned( editor )
+
+		isFollowingNodeSelection = drivingEditor is None and not isPinned
+		m.append( "/Following Selected Nodes", {
```
How about:

`m.append( "/Follow Node Selection", {`

I would stick to the same phrasing you used elsewhere, to avoid ambiguity where the user might think they can focus on more than one node at a time for editors that can't yet.

I'm not completely sold on "following" as being the correct verb here. Was "focus" avoided for a reason?

themissingcow replied:
Yeah, we're not entirely happy with that whole thing.

>   * I'm not sure about distinguishing solely by colour on such a small element, when colour blindness and visual acuity of aging artists is something to consider.

Any ideas for alternatives?

>   * I'm also not sure 4 colours will be enough for some very messy graphs, especially now that we've opened the door to multi-monitor layouts. It wasn't clear to me at first that when I was linking to my 5th _Viewer_ that this was a _new_ red link, and not the first one again.

We can easily add more :) how many do you think is a realistic upper bound?

themissingcow replied:
> I would stick to the same phrasing you used elsewhere, to avoid ambiguity where the user might think they can focus on more than one node at a time for editors that can't yet.

Skipping the follow/focus change for a mo, as this applies to either, I think the thinking was that the top half is state, so it describes 'how it is'. Following... Pinned... If we were to change it to 'Follow' we'd also have to change it to 'Pin'. Then when you have another item below, that is purely a one-shot action, 'Pin ...' is the different between the two the apparent? 

Youd end up with:

```
Follow Selected Nodes
Pin
Link to Editor >
---------
Pin Current Node Selection
Pin Bookmark
--------
Bookmark <current node>
```

medubelko replied:
> Any ideas for alternatives?

The only thing that came to mind was to use a number next to the pin icon, which follows the pattern we established with bookmarked nodes. I suppose there's a potential to confuse editor number with bookmark number, but at least it will be consistent? This also solves the need to define a finite number of colours, and accommodate various people's eyesight.

> Skipping the follow/focus change for a mo...

Oops, I actually wanted to avoid splitting hairs on the infinitive/imperative distinction, but I wrote `Follow` without thinking. Freudian typing, I suppose.

I think using imperative everywhere is fine. I know it's inconsistent with other "state" menu items we've established, but those are already inconsistent, with actions like _Show Input/Output Connections_ in the node right-click menu.

But I don't want to derail you, so let's talk about UI text conventions elsewhere. For this PR, I'm more interested in "focusing" vs. "following".

johnhaddon replied:
> The only thing that came to mind was to use a number next to the pin icon, which follows the pattern we established with bookmarked nodes.

The bookmark connotation you highlighted is problematic here I think. We plan to extend the pinning options with "Following Bookmark 1" etc, so any use of numbers should be reserved to that. My only other suggestion would be shape, but I think colour is probably a better compromise.

> I'm not completely sold on "following" as being the correct verb here. Was "focus" avoided for a reason?

Following implies continuous activity - i.e. the editor will adjust to follow future changes. Focussing implies a temporary transition state on the way to being focussed on one thing. We also have a need for the reverse term in tooltips : "Followed by ...", and there doesn't seem to be a good phrasing of that for "focus".

> I actually wanted to avoid splitting hairs on the infinitive/imperative distinction

I'm actually quite interested in this one, but as you say, let's discuss at leisure and looking at the UI as a whole, rather than focus on it for just this PR.

medubelko commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1278,3 +1376,96 @@ def _reprDict( d ) :
+			toolTipElements.append( "Following _%s_" % masterEditor.getTitle() )
+
+		if drivenEditors :
+			toolTipElements.append( "Folllowed by:")
```
`toolTipElements.append( "Followed by:")`

themissingcow replied:
Fixed in https://github.com/GafferHQ/gaffer/commit/aca4e28c8ebb8804581443491d1e9f96db3798b5

medubelko commented on startup/gui/editor.py:
```diff
@@ -44,8 +45,106 @@ def __editorCreated( editor ) :
+		if target is not editor :
+			__addFollowMenuItem( menuDefinition,
+				editor, target,
+				"Following Scene Selection",
```
`"Focusing on scene selection",` ?

themissingcow commented:
Hey, thanks for the input, good thoughts indeed.

> * Is there a reason I can't have the _Viewer_ follow another editor? It's not something I would do, but it seems like an arbitrary limitation.

We found that listing all combinations was somewhat overwhelming. We can certainly open it up for the general follow mode if people wish.

For the scene selection modes, the `Viewer` and `Hierarchy View` are the only two that allow you to manipulate the scene selection, so it felt it only really make sense in conjunction with one of those.

>   * Since the first coloured dot I see is red, I instinctively think of "recording," with no connection to the link.

Re-ordered the colors for now.

>   * I'm not sure about distinguishing solely by colour on such a small element, when colour blindness and visual acuity of aging artists is something to consider.

Yeah, we're hoping to refine this over the coming releases.

>   * I'm also not sure 4 colours will be enough for some very messy graphs, especially now that we've opened the door to multi-monitor layouts. It wasn't clear to me at first that when I was linking to my 5th _Viewer_ that this was a _new_ red link, and not the first one again.

I've added another color for now, and we'll address this properly when we adjust this whole presentation more generally.

johnhaddon commented:
Hey Tom, I took a look at fixing those conflicts, but I'm not super-confident in my svg-diff-comprehension. Might be better if you fixed it up on Monday, knowing what the intended end result is :)
